### PR TITLE
[HCP Observability] OTELExporter

### PR DESCRIPTION
**Pull request by Achooo**
##### Overall Feature background
This PR is part of the HCP Observability feature to send Server metrics to HCP's new Telemetry Gateway. The idea is to create a new OpenTelemetry go-metrics sink that will be initialized automatically when a Consul server is registered with the HCP management plane.
- We will be merging these approved changes into **feature/hcp-telemetry**
- That branch will be merged at the end into `main` and back ported
This allows us to iteratively improve on this feature over time before. Split up for easier PR review. 🚢 

##### This PR: MetricsClient

In #17100 , we added a MetricsClient capable of HCP auth to export metrics.

This PR adds a new `OTELExporter` to the new Telemetry package.
It must satisfy the [OTEL Metrics SDK Exporter](https://github.com/open-telemetry/opentelemetry-go/blob/exporters/otlp/otlpmetric/otlpmetrichttp/v0.37.0/sdk/metric/exporter.go) interface to be used by the OTEL go-metrics sink (follow up PR).
The key responsibility of the OTELExpoter is to use the MetricsClient to export metrics. To do this:
- Converts OTEL data format -> OTLP data format
- Uses the Client to perform the export HTTP request

##### Next steps
The OTELSink will use this exporter to periodically export metrics via the PeriodicReader.

##### Reference image

minirfc_optionA_exporter drawio-6

### PR Checklist

* [x] updated test coverage
Screenshot 2023-04-25 at 12 32 30 PM

* [ ] external facing docs updated
* [ ] appropriate backport labels added (not yet, in feature branch)
* [ ] not a security concern

### CODE REVIEW

File: agent/hcp/telemetry/otlp_transform.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
+package telemetry
```
Comment (Achooo) [inline agent/hcp/telemetry/otlp_transform.go:1]: I was hoping to re-use the implementation by the opentelemetry-go library for the conversion between the OTEL metrics sdk to the OTLP proto.

Turns out that [package](https://github.com/open-telemetry/opentelemetry-go/tree/exporters/otlp/otlpmetric/otlpmetrichttp/v0.37.0/exporters/otlp/otlpmetric/internal/transform) is an internal one. Here's how it's [used](https://github.com/open-telemetry/opentelemetry-go/blob/2e54fbb3fede5b54f316b3a08eab236febd854e0/exporters/otlp/otlpmetric/internal/exporter.go#L54).

I used the same implementation but I was able to reduce the code for our use-case to the current file.

File: agent/hcp/telemetry/otel_exporter.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// Temporality returns the Cumulative temporality for metrics aggregation.
+// Telemetry Gateway stores metrics in Prometheus format, so use Cummulative aggregation as default.
+func (e *OTELExporter) Temporality(_ metric.InstrumentKind) metricdata.Temporality {
+	return metricdata.CumulativeTemporality
+}
+
+// Aggregation returns the Aggregation to use for an instrument kind.
+func (e *OTELExporter) Aggregation(kind metric.InstrumentKind) aggregation.Aggregation {
```
Comment (Achooo) [inline agent/hcp/telemetry/otel_exporter.go:28]: We could use the default provided by the library to make this function a one-liner :  `return metric.DefaultAggregationSelector(kind)`

However, this implementation [panics](https://github.com/open-telemetry/opentelemetry-go/blob/2e54fbb3fede5b54f316b3a08eab236febd854e0/sdk/metric/reader.go#L147), which worries me a bit. So I wrote a custom simple one.

File: agent/hcp/telemetry/otel_exporter.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	switch kind {
+	case metric.InstrumentKindObservableGauge:
+		return aggregation.LastValue{}
+	case metric.InstrumentKindHistogram:
+		return aggregation.ExplicitBucketHistogram{
+			Boundaries: []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000},
```
Comment (Achooo) [inline agent/hcp/telemetry/otel_exporter.go:34]: I just used the same default as the implementation linked in above comment. I'm not sure yet what boundaries we want to set, this might also be dependent on the metric. CC @nickethier

**Review by loshz — CHANGES_REQUESTED** [review]
Overall, I'm happy with this implementation! This is really great work, Ash.

Added a couple of comments/nits.

File: agent/hcp/telemetry/otel_exporter.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
+package telemetry
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter.go:1]: Let's add a doc.go with a package description. @clly  does this in the collector and it's quite nice.

File: agent/hcp/telemetry/otel_exporter.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	"go.opentelemetry.io/otel/sdk/metric/aggregation"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+)
+
+// OTELExporter is a custom implementation of a OTEL Metrics SDK metrics.Exporter.
+// The exporter is used by a OTEL Metrics SDK PeriodicReader to export aggregated metrics.
+// This allows us to use a custom client - HCP authenticated MetricsClient.
+type OTELExporter struct {
+	client   hcpclient.MetricsClient
+	endpoint string
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter.go:19]: Chappie: Maybe use a url.URL

File: agent/hcp/telemetry/otel_exporter.go (hunk 3 above)
Comment (chapmanc) [reply agent/hcp/telemetry/otel_exporter.go:43]: Let's follow up on this and add a ticket to measure this after we have some metrics incoming.

File: agent/hcp/telemetry/otel_exporter.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			NoMinMax:   false,
+		}
+	}
+	// for metric.InstrumentKindCounter and others, default to sum.
+	return aggregation.Sum{}
+}
+
+// Export serializes and transmits metric data to a receiver.
+func (e *OTELExporter) Export(ctx context.Context, metrics metricdata.ResourceMetrics) error {
+	if len(metrics.ScopeMetrics) == 0 || len(metrics.ScopeMetrics[0].Metrics) == 0 {
```
[the remark this answers is not in the document]
Comment (chapmanc) [reply agent/hcp/telemetry/otel_exporter.go:53]: Chappie: sometimes naming conditions outside of the if block gives reviewers/code viewers a better understanding of the code.


```
// emtpyMetrics is occuring through the otel code due to .. 
emtpyMetrics := len(metrics.ScopeMetrics) == 0 || len(metrics.ScopeMetrics[0].Metrics) == 0

if emtpyMetrics {

...
```

Note: Ash mentioned in an sync review of this code that we don't control the call to export and the sdk is gathering metrics and calling this. In some circumstances it's sending an empty object through when there has been no change (theory right now).

File: agent/hcp/telemetry/otel_exporter.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+		return nil
+	}
+
+	otlpMetrics, merr := transformOTLP(&metrics)
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter.go:57]: You _might_ be able to just use the proto unmarshaller but let's rule it out. I'd rather us not have to use our own crafted converters if possible.

Comment (chapmanc) [reply agent/hcp/telemetry/otel_exporter.go:57]: We might want to check this merr before we export Metrics? Preferably we should return known error types vs mutlierrors so we can switch and act on the type.

File: agent/hcp/telemetry/otel_exporter_test.go [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+	"go.opentelemetry.io/otel/sdk/resource"
+	metricpb "go.opentelemetry.io/proto/otlp/metrics/v1"
+)
+
+func TestTemporality(t *testing.T) {
+	exp := &OTELExporter{}
+	require.Equal(t, metricdata.CumulativeTemporality, exp.Temporality(metric.InstrumentKindCounter))
+}
+
+func TestAggregation(t *testing.T) {
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter_test.go:23]:
```suggestion
func TestAggregation(t *testing.T) {
t.Parallel()
```

File: agent/hcp/telemetry/otel_exporter_test.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+			kind:   metric.InstrumentKindCounter,
+			expAgg: aggregation.Sum{},
+		},
+		"histogram": {
+			kind:   metric.InstrumentKindHistogram,
+			expAgg: aggregation.ExplicitBucketHistogram{Boundaries: []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 750, 1000, 2500, 5000, 7500, 10000}, NoMinMax: false},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			exp := &OTELExporter{}
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter_test.go:42]:
```suggestion
	} {
	        test = test
		t.Run(name, func(t *testing.T) {
		        t.Parallel()
			exp := &OTELExporter{}
```

File: agent/hcp/telemetry/otel_exporter_test.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+				Resource: resource.Empty(),
+				ScopeMetrics: []metricdata.ScopeMetrics{
+					{
+						Metrics: []metricdata.Metrics{
+							{
+								Name: "consul.raft.commitTime",
+							},
+						},
+					},
+				},
```
Comment (chapmanc) [inline agent/hcp/telemetry/otel_exporter_test.go:94]: Might be easier to read if we have a `mutateMetrics(m []metricdata.ScopeMetrics)` to modify per testcase

File: agent/hcp/telemetry/otlp_transform.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+		ScopeMetrics: sms,
+	}, err
+}
+
+// scopeMetrics returns a slice of OTLP ScopeMetrics.
+func scopeMetrics(scopeMetrics []metricdata.ScopeMetrics) ([]*mpb.ScopeMetrics, error) {
+	var merr error
+	out := make([]*mpb.ScopeMetrics, 0, len(scopeMetrics))
+	for _, sm := range scopeMetrics {
+		ms, err := metrics(sm.Metrics)
```
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform.go:32]: metricsToProto?

File: agent/hcp/telemetry/otlp_transform_test.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+		TimeUnixNano:      uint64(end.UnixNano()),
+		Count:             3,
+		Sum:               &sumB,
+		ExplicitBounds:    []float64{1, 5},
+		BucketCounts:      []uint64{0, 1, 2},
+		Min:               &minB,
+		Max:               &maxB,
+	}}
+	// DataPoint test case : Number Datapoints (Gauge / Counter)
+	otelDP = []metricdata.DataPoint[float64]{
```
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform_test.go:82]: input?

File: agent/hcp/telemetry/otlp_transform_test.go [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+		{Attributes: alice, StartTime: start, Time: end, Value: 1.0},
+		{Attributes: bob, StartTime: start, Time: end, Value: 2.0},
+	}
+
+	otlpDP = []*mpb.NumberDataPoint{
```
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform_test.go:87]: expectedOutput?

File: agent/hcp/telemetry/otlp_transform_test.go [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+				},
+			},
+		},
+		ScopeMetrics: otlpScopeMetrics,
+	}
+)
+
+// TestTransformOTLP runs tests from the "bottom-up" of the metricdata data types.
+func TestTransformOTLP(t *testing.T) {
+	// Histogram DataPoint Test Case (Histograms)
```
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform_test.go:237]: t.Parallel?

File: agent/hcp/telemetry/otlp_transform_test.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+	require.Equal(t, otlpDP, dataPoints(otelDP))
+
+	// Metrics Test Case
+	m, err := metrics(otelMetrics)
+	require.Equal(t, otlpMetrics, m)
+	require.Equal(t, len(otlpMetrics), 3)
+	require.Contains(t, err.Error(), "3 errors occurred")
+
+	// Scope Metrics Test Case
+	sm, err := scopeMetrics(otelScopeMetrics)
```
Comment (chapmanc) [inline agent/hcp/telemetry/otlp_transform_test.go:250]: If we can name these functions different from the struct name it might help to be a bit more descriptive as to what they do. ex: `scopeMetricsToPB`

File: go.mod [hunk 16]
```diff
@@ -96,13 +96,16 @@ require (
 	github.com/shirou/gopsutil/v3 v3.22.8
 	github.com/stretchr/testify v1.8.2
 	go.etcd.io/bbolt v1.3.6
+	go.opentelemetry.io/otel v1.15.0-rc.1
```
Comment (chapmanc) [inline go.mod:99]: We have to use the rc because there are import errors that are fixed in the RC. Link? Can we add a todo/jira to follow up once we have a stable release.

File: agent/hcp/telemetry/otel_exporter.go (hunk 4 above)
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go:1]: Done - would love feedback on the writeup.

File: agent/hcp/telemetry/otel_exporter.go (hunk 2 above)
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go:40]: done! thanks

File: go.mod (hunk 16 above)
Comment (Achooo) [reply go.mod:99]: Actually it got released https://github.com/open-telemetry/opentelemetry-go/releases/tag/v1.15.1
🥳 So I'll update this!

File: agent/hcp/telemetry/otel_exporter.go (hunk 3 above)
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go:46]: Done! Added to GA epic for now but potentially could be picked up before

File: agent/hcp/telemetry/otel_exporter_test.go [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+func (m *mockErrMetricsClient) ExportMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics, endpoint string) error {
+	return fmt.Errorf("failed to export metrics")
+}
+
+type mockMetricsClient struct{}
+
+func (m *mockMetricsClient) ExportMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics, endpoint string) error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter_test.go]: yep made it much nicer thanks!

File: agent/hcp/telemetry/otlp_transform_test.go (hunk 15 above)
Comment (Achooo) [reply agent/hcp/telemetry/otlp_transform_test.go:250]: yep done!

File: agent/hcp/telemetry/otlp_transform.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+		if err != nil {
+			merr = multierror.Append(merr, err)
```
[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otlp_transform.go]: Yeah so we don't want to prevent appending (as we want the good metrics in) but we want to filter bad metrics (which is done in the inner function). I've cleaned up this code to remove the error handling here since it's pretty much useless. After discussing with Chappie and Connor, we'd rather get metrics on when transformation failed, and we'd do this within the inner functions when the error occurs. I'm thinking of doing this in a follow-up PR. The alternative is to log the error, but it would be easier to obtain metrics.

File: agent/hcp/telemetry/otel_exporter.go (hunk 7 above)
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go]: I've updated this to remove that error, since we plan on sending a metric for errors

File: agent/hcp/telemetry/otel_exporter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go]: I could iterate through all ScopeMetrics and all Metrics within (since the transform is generic). That would make this code more extensible in case someone uses this with multiple Meters (so multiple ScopeMetrics). I think that would address your concern @loshz ?

[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go]: Added a new `isEmpty()` method that checks if there are any metrics to be sent

File: agent/hcp/telemetry/doc.go [hunk 19]
```diff
@@ -0,0 +1,12 @@
+// Package telemetry implements functionality to collect, aggregate, convert and export
+// telemetry data in OpenTelemetry Protocol (OTLP) format.
+//
+// The entrypoint is the OpenTelemetry (OTEL) go-metrics sink which:
```
Comment (Achooo) [inline agent/hcp/telemetry/doc.go:4]: Coming up in follow up PR but added comment now anyways

File: agent/hcp/telemetry/otel_exporter.go [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+	return aggregation.Sum{}
+}
+
+// Export serializes and transmits metric data to a receiver.
+func (e *OTELExporter) Export(ctx context.Context, metrics metricdata.ResourceMetrics) error {
+	otlpMetrics := transformOTLP(&metrics)
+	if isEmpty(otlpMetrics) {
+		return nil
+	}
+	return e.client.ExportMetrics(ctx, otlpMetrics, e.url.String())
```
Comment (Achooo) [inline agent/hcp/telemetry/otel_exporter.go:59]: Should the client also be changed to handle a URL instead of a string endpoint? CC @chapmanc @loshz

File: agent/hcp/telemetry/otel_exporter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (loshz) [reply agent/hcp/telemetry/otel_exporter.go]: Yep, this makes sense. I guess I'm just unsure why we use a slice if we know there is only ever 1 item? I assume this is coming directly from he SDK?

File: agent/hcp/telemetry/otlp_transform.go [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+				Version: sm.Scope.Version,
+			},
+			Metrics: ms,
+		})
+	}
+	return out, merr
+}
+
+// metrics returns a slice of OTLP Metric generated from OTEL metrics sdk ones.
+func metrics(metrics []metricdata.Metrics) ([]*mpb.Metric, error) {
```
[the remark this answers is not in the document]
Comment (loshz) [reply agent/hcp/telemetry/otlp_transform.go]: @Achooo this is marked as resolved but I don't see what changed?

**Comment by loshz** [comment]
@Achooo thanks for addressing some of the PR comments! The code is looking great. Let me know when you'd like me to re-review.

File: go.mod (hunk 16 above)
Comment (Achooo) [reply go.mod]: Updated!

File: agent/hcp/telemetry/otel_exporter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go]: Yeah it's unfortunately the output of the SDK after aggregating, given to the exporter 😓

File: agent/hcp/telemetry/otel_exporter.go (hunk 7 above)
Comment (Achooo) [reply agent/hcp/telemetry/otel_exporter.go]: Re - proto unmarshaleller

I'd like to keep the current implementation, rationale:

1. I saw that the Gateway uses a [ProtoUnmarshaler](https://pkg.go.dev/go.opentelemetry.io/collector/pdata@v0.66.0/pmetric#ProtoUnmarshaler) provided by the pdata (internal representation of metrics). However, I don't see support for marshal/unmarshal in the metrics SDK [here](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric@v0.38.1/metricdata#) and protobufs [here](https://pkg.go.dev/go.opentelemetry.io/proto/otlp@v0.19.0/metrics/v1). 
2. As we discussed, it might be a good idea to have this to prevent a single metric transformation error from sending all metrics

File: agent/hcp/telemetry/otlp_transform.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (Achooo) [reply agent/hcp/telemetry/otlp_transform.go]: Woops my bad! This was also brought up by Chappie, and I think I meant to tag you there, see this [thread](https://github.com/hashicorp/consul/pull/17128#discussion_r1187757389).

TLDR: The library doesn't provide anything :( and I wish it did. As mentioned in this [comment](https://github.com/hashicorp/consul/pull/17128#discussion_r1176791223), the conversion package is also internal and there are no marshal/unmarshall or ToProto() functions provided to make this easier.

**Comment by Achooo** [comment]
@chapmanc I'm working on a follow-up PR with Consul metrics for metrics in this package. The code seems very straightforward but I'd like to test it locally, also to unblock this PR for now.

**Review by loshz — APPROVED** [review]
LGTM - thanks for taking the time to address the comments!
